### DISCUSSION TITLE
Using MATLAB to run SimNIBS and create visualisations

### DISCUSSION THREAD

**Discussion by RachelETHOMAS** [simnibs/simnibs#159] [2. General]
I am trying to run SimNIBS scripts in MATLAB and when I do, as well as with the visualisation commands, an error message comes up saying that  'Dot indexing is not supported for variables of this type', and I can't seem to get any further around this.

e.g., 
I'm able to run the command:
mesh_get_simulation_result

but I am then unable to correctly use:
mesh_show_surface

I'm using version 2021 of MATLAB- maybe it's related to this?

If there's any advice you could give me that would be great!

**Comment by RachelETHOMAS** [comment]
Myself and my supervisor, have got the first part of the Scipting Simulations code to run (setting the paths up in Matlab), but we're unable to get the "run_simnibs(s)" to work. 

If we isolate it by commenting it out from the script and running it after, the first part of the script works fine to create Variable. But after this 
point we encounter issues. 

"run_simnibs.m" gets further in an older version of simnibs (version 3.2), it'll create the "tutorial" folder, but again just throw up errors in a log file within that tutorial folder.

We've tried taking the run_simnibs.py script and making a modification to include absolute system paths to the simnibs software. But again this has been to no avail.

Pasted below is the error message from within Matlab 2021a after running "run_simnibs(s)":

**ImportError:
/cubric/software/simnibs.version/4.0.0/simnibs_env/lib/python3.9/site-packages/simnibs/mesh_tools/cgal/create_mesh_surf.cpython-39-x86_64-linux-gnu.so:
undefined symbol: __gmpq_add**

Error using run_simnibs (line 30)
There was an error running SimNIBS:
Traceback (most recent call last):
File "(/cubric/software/simnibs.versions/4.0.0/simnibs_env/lib/python3.9/site-packages/simnibs/cli/run_simnibs.py", line 7, in <module> from simnibs import run_simnibs
File "(/cubric/software/simnibs.versions/4.0.0/simnibs_env/lib/python3.9/site-packages/simnibs/__init__.py", line 21, in <module> from .mesh_tools import *
File "(/cubric/software/simnibs.versions/4.0.0/simnibs_env/lib/python3.9/site-packages/simnibs/mesh_tools/__init__.py", line 6, in <module> from .mesh_io import *
File "(/cubric/software/simnibs.versions/4.0.0/simnibs_env/lib/python3.9/site-packages/simnibs/mesh_tools/mesh_io.py", line 49, in <module> from .import cgal
File "(/cubric/software/simnibs.versions/4.0.0/simnibs_env/lib/python3.9/site-packages/simnibs/mesh_tools/cgal/__init__.py", line 1, in <module> from .create_mesh_surf import*

**Comment by jduemose** [comment]
Hi, given that the import error stems from the shared object (.so) file, I would imagine that this is a linking issue. We have seen this before when calling simnibs functions from MATLAB.

To confirm this, you could try to run `unix("ldd /cubric/software/simnibs.version/4.0.0/simnibs_env/lib/python3.9/site-packages/simnibs/mesh_tools/cgal/create_mesh_surf.cpython-39-x86_64-linux-gnu.so")` from within MATLAB. This shows the libraries against which the this object is linked. The libgmp* link should point inside the simnibs folder. If not, this is likely the issue.

To resolve this, you could try to prepend the libraries shipped with simnibs to the `LD_LIBRARY_PATH` like so

`setenv('LD_LIBRARY_PATH', sprintf('/cubric/software/simnibs.version/4.0.0/simnibs_env/lib/python3.9/site-packages/simnibs/external/lib/linux:%s', getenv('LD_LIBRARY_PATH')))`

Alternatively, using a similar strategy, you can set `LD_PRELOAD` and to point to the specific libraries (.so files) in this folder.

Hope this helps,

Jesper

**Comment by RachelETHOMAS** [comment]
Hi Jesper,

That was great thanks. While that exact solution didn't work for us, we fixed the problem by adding the following to the top of the script:

%Set Matlab env variable for libgmp.so.3
setenv('LD_LIBRARY_PATH', '/cubric/software/simnibs.versions/4.0.0/simnibs_env/lib/python3.9/site-packages/simnibs/external/lib/linux/libgmp.so.3')
